Fixes PTV-1694 - search input not retained when switching detail tabs, fetching more results, or primary tabs

eapearson opened the pull request:
This bug was reported  in public ticket https://kbase-jira.atlassian.net/browse/PUBLIC-1687 and tracking ticket https://kbase-jira.atlassian.net/browse/PTV-1694.

The initial problem was trivial - the 'search' url param was simply not retained during navigation actions for the affected tabs and buttons.

Some changes were required to build the app, and more to get tests to pass.

Dependencies needed to be pinned and updated to stabilize the codebase. More work in that area is needed, because some libraries are a major-version out of date and require code changes to accommodate.

Added a changelog document.

Updated to node16 in the image - was based on node10 which is out of support. node16 is an LTS release with several years left of support.

Prettier update required a round of prettier-formatting.

Some tests were passing, but actually triggered errors which were buried in console logging. One such fix, of the org association component, resulted in a refactor to separate out an embedded component and data access code in order to help isolate the problem.

In another "tag along" change, language was aligned with the branding of this product as "Narratives Navigator". This resulted in the title becoming "Narratives Navigator", the menu label becoming "Navigator", and the icon becoming a compass.

eapearson committed:
first pass at updating and pinning dependencies

eapearson committed:
prettier-reformatted files (after updating prettier version, which was very out of date)

eapearson committed:
move to node 16, node 10 is out of support

eapearson committed:
need -n for entr to work (crashes otherwise)

eapearson committed:
ignore virtual env

eapearson committed:
fix TS usage

eapearson committed:
proxy should default to ci; it is the "ci proxy" after all!

eapearson committed:
preserve 'search' param when switching between tabs Data and Preview

eapearson committed:
preserve search when clicking the "Load more..." button

eapearson committed:
improve "no data" message

eapearson committed:
remove console log

eapearson committed:
add key to stop react warning

eapearson committed:
loading spinner test should not assume implemented as <i> - it doesn't matter; improve tests

eapearson committed:
remove console log lines

eapearson committed:
refactor OrgSelect out of LinkOrgItem; add basic test

eapearson committed:
refactor LinkOrgItem to extract Org selector, extract some data access code, improve async process management

eapearson committed:
period should be colon; remove unnecessary comments and es-lint directives; add mock to fix test

eapearson committed:
fix test mock

eapearson committed:
Add typing (should have very sparing but preferrably no use of "any")

eapearson committed:
improve warning, and add TODO comment about future improvement

eapearson committed:
update license to 2021

eapearson committed:
update and pin dependencies; fix TS issues mostly exception value usage (TS was unpinned, and eventually was triggered to update to latest)

eapearson committed:
add changelog document and initial comments

eapearson committed:
update readme, add macos instructions

eapearson committed:
lock file

eapearson committed:
fix typing

eapearson committed:
retain search between tabs

eapearson committed:
fix navigator menu item: should be "Navigator" and compass icon; should be highlighted; adjust spacing to match kbase-ui

eapearson committed:
Update to use "official" name for this - Narratives Navigator

eapearson committed:
use braces for if statement

eapearson committed:
simplify conditional default value logic

eapearson committed:
use braces for if statement

eapearson committed:
comment note on fontawesome version; really should provided in codebase via node dependency.

eapearson committed:
improve test language and add a TODO

eapearson committed:
static strings don't need to be evaluated; add some

eapearson committed:
Update changelog

eapearson committed:
remove unused import and unnecessary comments

eapearson committed:
eslint-plugin-react was missing

eapearson committed:
ignore jetbrains ide config

eapearson committed:
eslint prettier config no longer requires individual prettier plugin extends

eapearson committed:
remove unused variable

dakotablair commented on src/templates/narratives/index.html:
```diff
@@ -1,11 +1,15 @@
-{% extends template %}
+{% extends template %} 
```
There is some more trailing whitespace in this file.

eapearson replied:
Now end-of-line spaces should be gone from this file.

dakotablair commented on src/client/api/iconProvider.ts:
```diff
@@ -60,7 +60,7 @@ export default class IconProvider {
 
   private processLoadedTypes() {
-    Object.keys(ICON_DATA.data).forEach(t => {
+    Object.keys(ICON_DATA.data).forEach((t) => {
```
My linter complains about single variables inside of parentheses, your prettier settings seem to go in the opposite direction. We might need to confer to make the environment consistent in different dev environments. Similarly, running `yarn fix` undoes all these changes for me.

eapearson replied:
Strange, if I run `yarn fix` it actually reformats to wrap function parameters in parens. 
I think I noted in the PR description that prettier was out of date, and that some of the prettier-formatting diffs were probably due to this.

eapearson replied:
Ideally all files affected by a linter or linter rule change would be modified and included in a changeset.

dakotablair replied:
Yes, we should probably get together after the meeting to make sure that `yarn fix` is idempotent in both of our environments. I imagine there is some configuration that can be included or modified in the repo that will fix the issue.

eapearson replied:
After the latest commit and a fresh clone of this branch, `yarn fix` does not result in any changes.

dakotablair reviewed (changes requested):
Erik, here are a few little things, but we should probably discuss the linter/prettier issues.

eapearson commented on Dockerfile:
```diff
@@ -1,5 +1,5 @@
 # For installing npm dependencies and building static css/js
-FROM node:10-alpine
+FROM node:16-alpine3.14
```
True that 16 is not on LTS yet, it will be. So 16 is most recent release, intended for LTS, with the longest support lifespan. And this is only used a build tool, so honestly the version matters less, what matters most is that it is supported and that we won't have to change it for a while.

14 is the current LTS, but will soon be in maintenance. It could be good to be on a maintenance release, for reduced risk. For us, this is just for development and build, so I don't think the risks are high. Choosing a later release will just move the goalposts further.

Good suggestion on the [engines].

eapearson replied:
Added engines selecting node 16, and also `engine-strict`, which makes the package.json compatible with npm.
`yarn` 1 only has `engine-strict` enabled if engines are used. `yarn` 2 relaxes this; I don't know if it uses `engine-strict`, but probably does as this caused problems for some users.

This change is nice for consistency, as it ensures that only node 16 will always be used with the project.

We don't use node-16-only features ourselves afaik, but I've found that some libraries can come to depend on features of more recent node releases, and are often tested only with more recent releases. Certainly our previous usage of node 10 would be problematic, as it was out of support and I doubt many libraries test against it.

dakotablair replied:
Bah, now github actions refuses to run node 16.
https://github.com/kbaseIncubator/dashboard-redesign/pull/88/checks?check_run_id=3478619705#step:3:6

I think that has to do with this line in the github action definition:
https://github.com/kbaseIncubator/dashboard-redesign/blob/main/.github/workflows/test.yml#L11

The following link indicates that, today, `ubuntu-latest` refers to 20.04:
https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idruns-on

So we need to either use >=14.17.5 for now or update node in the github action. If it were me, I would move to 14.17.5 and update the github `runs-on` to be `ubuntu-20.04`. I'm not sure if it is better to pin the engine to 14.17.5 (since that's what the CI will always be running on) or make it `>=` to be more flexible.

dakotablair replied:
Nice, I was not sure how easy it would be to specify the node version. Fortunately it was easy! Thanks 👍

eapearson committed:
markdown linting fixes

eapearson commented on CHANGELOG.md:
```diff
@@ -0,0 +1,37 @@
+
+### Fixed
+
+- Fix disappearing search text when use the Data/Preview tab, or the "Load More" button 
```
Also an empty line after the title. Markdown linter installed, commit pushed

eapearson commented on dev/nginx-default.conf:
```diff
@@ -22,7 +22,7 @@ server {
          add_header 'Access-Control-Expose-Headers' 'Authorization,DNT,X-CustomHeader,Keep-Alive,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Content-Range,Range' always;
          proxy_hide_header 'Access-Control-Allow-Origin';
-         proxy_pass https://kbase.us;
+         proxy_pass https://ci.kbase.us;
```
Yes, by default development should be against CI not prod. There are also several other bits of development support that either encode or document working against ci.

eapearson committed:
changelog lint fix

eapearson committed:
remove spaces from ends of lines

eapearson committed:
run prettier over entire codebase - these are all the changes including (including the eslint config which also has an edit). The main formatting change is to enforce parens around fat-arrow function paramters, other than a couple of files which must have been missed in earlier prettier runs (or are now being reformatted under prettier 2.) Also remove arrow-parens rule from eslint - it conflicts with prettier (and the google plugin already does what we want to do but is redundant, trigger eslint to require arrow-parens)

eapearson committed:
add engine node 16 (only), engine-strict to be compatible with npm
Note that yarn 1 is strict about the specified engine; it will fail to install if a node runtime other than 16 is used.

eapearson committed:
add node version to gha

dakotablair commented:
This PR really snuck in some other needed updates next to your change, but the project is better for it! Thanks again.
